apps/web/src/components/templates/TemplateEditorProvider.tsx

```diff
@@ -1,132 +1,315 @@
-    });
-  };
-
-  const setIsEmbedModalVisible = (isEmbedModalVisible: boolean) => {
-    dispatch({
-      type: actionIsEmbedModalVisible,
-      payload: isEmbedModalVisible,
-    });
-  };
-
-  const setTrigger = (trigger: INotificationTrigger) => {
-    dispatch({
-      type: actionTrigger,
-      payload: trigger,
-    });
-  };
-
-  const setCreatedTemplateId = (createdTemplateId: string) => {
-    dispatch({
-      type: actionCreatedTemplateId,
-      payload: createdTemplateId,
-    });
-  };
-
-  return (
-    <Provider
-      value={{
-        setTrigger,
-        setIsDirty,
-        setEditMode,
-        setIsEmbedModalVisible,
-        setCreatedTemplateId,
-        editMode: state.editMode,
-        isDirty: state.isDirty,
-        isEmbedModalVisible: state.isEmbedModalVisible,
-        trigger: state.trigger,
-        createdTemplateId: state.createdTemplateId,
-      }}
-    >
-      {children}
-    </Provider>
+  const setTrigger = useCallback(
+    (newTrigger: INotificationTrigger) => setState((old) => ({ ...old, trigger: newTrigger })),
+    [setState]
+  );
+
+  const setCreatedTemplateId = useCallback(
+    (newCreatedTemplateId: string) => setState((old) => ({ ...old, createdTemplateId: newCreatedTemplateId })),
+    [setState]
+  );
+
+  const {
+    reset,
+    formState: { isDirty: isDirtyForm },
+  } = methods;
+
+  const steps = useFieldArray({
+    control: methods.control,
+    name: 'steps',
+  });
+
+  const {
+    template,
+    isLoading,
+    isCreating,
+    isUpdating,
+    isDeleting,
+    updateNotificationTemplate,
+    createNotificationTemplate,
+  } = useTemplateController(templateId);
+
+  useEffect(() => {
+    if (isDirtyForm) {
+      return;
+    }
+
+    if (template && template.steps) {
+      const formValues: IForm = {
+        notificationGroupId: template._notificationGroupId,
+        name: template.name,
+        description: template.description ?? '',
+        tags: template.tags,
+        identifier: template.triggers[0].identifier,
+        critical: template.critical,
+        preferenceSettings: template.preferenceSettings,
+        steps: [],
+      };
+
+      formValues.steps = (template.steps as IStepEntity[]).map((item) => {
+        if (item.template.type === StepTypeEnum.EMAIL) {
+          return {
+            ...item,
+            ...(!item.replyCallback && {
+              replyCallback: {
+                active: false,
+                url: '',
+              },
+            }),
+            template: {
+              ...item.template,
+              layoutId: item.template._layoutId ?? '',
+              preheader: item.template.preheader ?? '',
+              content: item.template.content,
+              ...(item.template?.contentType === 'customHtml' && {
+                htmlContent: item.template.content as string,
+                content: [],
+              }),
+            },
+          };
+        }
+        if (item.template.type === StepTypeEnum.IN_APP) {
```

Comment:
Hopefully in the future we can add template editors for the other channels. With that in mind extracting these mappers by channel to separated functions I think would help when adding more. The `useEffect` is already meaty and would be even more.

Reply:
yes, totally makes sense to separate these into separate functions, will do that, thanks!